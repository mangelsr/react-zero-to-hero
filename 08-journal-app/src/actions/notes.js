import Swal from "sweetalert2";

import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


export const startNewNote = () => {
    return async (dispatch, getState) => {
        try {
            const { uid } = getState().auth;
            const newNote = {
                title: '',
                body: '',
                date: new Date().getTime()
            };
            const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
            dispatch(activateNote(doc.id, newNote));
            dispatch(addNewNote(doc.id, newNote));
        } catch (err) {
            console.log(err);
        }
    };
};

export const activateNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    };
};

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        try {
            const { uid } = getState().auth;
            if (!note.url) {
                delete note.url;
            }
            const noteToFirestore = { ...note };
            delete noteToFirestore.id;
            
            await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
            
            dispatch(refreshNote(note.id, note));
            
            Swal.fire('Note saved', note.title, 'success');
        } catch (err) {
            console.log(err);
        }
    };
};

export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: { id, note }
});

export const startFileUpload = (file) => {
    return async (dispatch, getState) => {
        try {
            const { active:activeNote } = getState().notes;
            Swal.fire({
                title: 'Uploading',
                text: 'Please wait...',
                allowOutsideClick: false,
                onBeforeOpen: () => Swal.showLoading()
            });
            const fileUrl = await fileUpload(file);
            Swal.close();
            activeNote.url = fileUrl;
            dispatch(startSaveNote(activeNote));
        } catch (err) {
            console.log(err);
        }
    };
};

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        try {
            const uid = getState().auth.uid;
            await db.doc(`${uid}/journal/notes/${id}`).delete();
            dispatch(deleteNote(id));
        } catch (err) {
            console.log(err);            
        }
    };
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const clearNotes = () => ({
    type: types.notesLogoutClean
});

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id, ...note
    }
});