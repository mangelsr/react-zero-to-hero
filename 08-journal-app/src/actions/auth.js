import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from "sweetalert2";

import { types } from "../types/types";
import { uiEndLoading, uiStartLoading } from "./ui";
import { clearNotes } from "./notes";

export const startLoginEmailPassword = (email, password) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithEmailAndPassword(email, password);
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithPopup(googleAuthProvider);
            dispatch(login(user.uid, user.displayName))
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};


export const startRegister = (email, password, name) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().createUserWithEmailAndPassword(email, password);
            await user.updateProfile({displayName: name});
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await firebase.auth().signOut();
            dispatch(logout());
            dispatch(clearNotes());
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        }
    };
};

export const logout = () => ({
    type: types.logout
});