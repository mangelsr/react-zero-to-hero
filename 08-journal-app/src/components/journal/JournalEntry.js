import React from 'react';
import moment from "moment";
import { useDispatch } from 'react-redux';
import { activateNote } from '../../actions/notes';


export const JournalEntry = ({ id, date, title, body, url }) => {
    
    const dispatch = useDispatch();
    
    const noteDate = moment(date);

    const handleActivateNote = () => {
        dispatch(activateNote(id, {
            title, body, date, url
        }));
    };

    return (
        <div onClick={handleActivateNote} className="journal__entry pointer animate__animated animate__fadeIn">
            {
                url &&
                <div
                    className="journal__entry-picture"
                    style={{
                        backgroundSize: 'cover',
                        backgroundImage: `url(${url})`
                    }}>
                </div>
            }
            <div className="journall__entry-body">
                <p className="journall__entry-title">
                    { title }
                </p>
                <p className="journall__entry-content">
                    { body }
                </p>
            </div>

            <div className="journall__entry-date-box">
                <span>{ noteDate.format('dddd') }</span>
                <h4>{ noteDate.format('Do') }</h4>
            </div>
        </div>
    );
}
