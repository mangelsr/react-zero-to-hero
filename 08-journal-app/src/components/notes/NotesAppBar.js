import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { startFileUpload, startSaveNote } from '../../actions/notes';


export const NotesAppBar = () => {
    
    const dispatch = useDispatch();
    const { active:note } = useSelector(state => state.notes);

    const handleSave = () => {
        dispatch(startSaveNote(note));
    };

    const handlePictureUpload = () => {
        document.querySelector("#fileSelector").click();
    };

    const handleFileChange = (e) => {
        const file = e.target.files[0];
        if (file) {
            dispatch(startFileUpload(file));
        }
    };
    
    return (
        <div className="notes__appbar">
            <span>28 de agosto 2020</span>
            <input id="fileSelector" type="file" accept="image/*" style={{display:'none'}} onChange={handleFileChange} />
            <div>
                <button onClick={handlePictureUpload} className="btn">
                    Picture
                </button>
                <button onClick={handleSave} className="btn">
                    Save
                </button>
            </div>
        </div>
    );
}
