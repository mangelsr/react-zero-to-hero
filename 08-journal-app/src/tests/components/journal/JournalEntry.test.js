import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';

import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();


const note = {
    id: 10,
    date: 0,
    title: 'Hello',
    body: 'world',
    url: 'https://someplace.com/img.jpg'
};

const wrapper = mount(
    <Provider store={store}>
        <JournalEntry {...note} />
    </Provider>
);

describe('Tests for JournalEntry component', () => {
   
    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should activate the note', () => {
        wrapper.find('div.journal__entry').simulate('click');
        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(note.id, note)
        );
    });
    
});