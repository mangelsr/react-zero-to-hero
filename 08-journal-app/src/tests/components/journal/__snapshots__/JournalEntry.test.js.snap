// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for JournalEntry component should render correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://someplace.com/img.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hello
    </p>
    <p
      className="journal__entry-content"
    >
      world
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
       
      Wednesday
       
    </span>
    <h4>
       
      31st
       
    </h4>
  </div>
</div>
`;
